function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,325 @@
+	InformFunctionReplicas(function FunctionId, replicas int)
+
+	// StartMonitoring starts monitoring metrics for the given topic and function.
+	StartMonitoring(topic string, function FunctionId) error
+
+	// StopMonitoring stops monitoring metrics for the given topic and function.
+	StopMonitoring(topic string, function FunctionId) error
+
+	// Propose proposes the number of replicas for functions that are being monitored.
+	Propose() map[FunctionId]int
+}
+
+// FunctionId identifies a function
+// TODO: support namespaces.
+type FunctionId struct {
+	Function string
+}
+
+// Go does not provide a MaxInt, so we have to calculate it.
+const MaxUint = ^uint(0)
+const MaxInt = int(MaxUint >> 1)
+
+// NewAutoScaler constructs an autoscaler instance using the given metrics receiver and the given transport inspector.
+func NewAutoScaler(metricsReceiver metrics.MetricsReceiver, transportInspector transport.Inspector) *autoScaler {
+	return &autoScaler{
+		mutex:               &sync.Mutex{},
+		metricsReceiver:     metricsReceiver,
+		transportInspector:  transportInspector,
+		totals:              make(map[string]map[FunctionId]*metricsTotals),
+		scalers:             make(map[FunctionId]scaler),
+		replicas:            make(map[FunctionId]int),
+		maxReplicas:         func(function FunctionId) int { return MaxInt },
+		delayScaleDown:      func(function FunctionId) time.Duration { return time.Duration(0) },
+		stop:                make(chan struct{}),
+		accumulatingStopped: make(chan struct{}),
+	}
+}
+
+func (a *autoScaler) SetMaxReplicasPolicy(maxReplicas func(function FunctionId) int) {
+	a.maxReplicas = maxReplicas
+}
+
+func (a *autoScaler) SetDelayScaleDownPolicy(delayScaleDown func(function FunctionId) time.Duration) {
+	a.delayScaleDown = delayScaleDown
+}
+
+func (a *autoScaler) Run() {
+	go a.receiveLoop()
+}
+
+type autoScaler struct {
+	mutex                      *sync.Mutex
+	metricsReceiver            metrics.MetricsReceiver
+	transportInspector         transport.Inspector
+	totals                     map[string]map[FunctionId]*metricsTotals
+	requiredScaleDownProposals int
+	scalers                    map[FunctionId]scaler
+	replicas                   map[FunctionId]int // tracks all functions, including those which are not being monitored
+	maxReplicas                func(function FunctionId) int
+	delayScaleDown             func(function FunctionId) time.Duration
+	stop                       chan struct{}
+	accumulatingStopped        chan struct{}
+}
+
+// metrics counts the number of messages transmitted to a Subscription's topic and received by the Subscription.
+type metricsTotals struct {
+	transmitCount int32
+	receiveCount  int32
+}
+
+func (a *autoScaler) Propose() map[FunctionId]int {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	proposals := make(map[FunctionId]int)
+	for _, funcTotals := range a.totals {
+		for fn, mt := range funcTotals {
+			proposals[fn] = a.scalers[fn](mt)
```

Comment:
If/when we support multi topic functions, a function may be registered to several topics and this overrides.
There should be aggregation logic (or at least fail fast so that we catch it when we introduce it)

Reply:
Nice catch! I _think_ the fix is to choose the maximum proposal for a function across all its topics, thus:
```
proposals[fn] = max(proposals[fn], a.scalers[fn](mt))
```

If you agree, I'll implement this and work out some tests.

Reply:
I think so too, but given this doesn't happen in practice I wonder if it isn't best to make it fail so that we make sure we revisit when the time comes (maybe with a comment hinting that your past-self thought `max()` was a good idea?)

Reply:
Good idea! It would be a waste of time to write tests for a "dead" scenario. I changed the code to panic (which should never happen until we start implementing multiple input topics, of course) and left a comment about `max()`.